### PR TITLE
CI monitor reading

### PR DESCRIPTION
**Pull request by raybejjani**
We see flakes in tests that read from monitor. These are structured to search monitor output for specific substrings. Since monitor may take some time to start up, we can change how we read its output. The 3 tests now read this incrementally, and fail on a timeout. This should allow it to pass when monitor was simply slow to print.

I haven't seen the specific flake in a bunch of testing with this PR. Most test runs have failed, however, often in other tests. I'm fairly confident that I didn't break anything but I would prefer more clean passes. It does do better on GKE and that is where we see the flake the most, however.

@joestringer I think we once discussed this flake in a PR, so I pulled you in for review. Feel free to unassign yourself.

### CODE REVIEW

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfig.*

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfig.*

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

File: test/k8sT/DatapathConfiguration.go
Comment (nebril) [inline test/k8sT/DatapathConfiguration.go]: won't this end up appending the same beginning of output over and over again, as we are not consuming the bytes that were already read. This may be ok if we just want to find "bad" lines, but I think it would be enough to just `monitorOutput = monitorRes.CombineOutput().Bytes()`. The same applies to all similar lines below (196, 890 in Policies.go).

File: test/k8sT/Policies.go
Comment (nebril) [inline test/k8sT/Policies.go]: We can drop the `sleep`s since we are using a timeout anyway (we can return `false` if exec is not successful)

**Review by nebril — COMMENTED** [review]
Two comments inline, good stuff overall

File: test/k8sT/DatapathConfiguration.go
Comment (raybejjani) [reply test/k8sT/DatapathConfiguration.go]: You're right. For some reason I thought that there was a .Read call on a bytes.Buffer somewhere but that doesn't seem to be the case. I've switch it out. A bit less code too! I realised that we already have a WaitUntilMatch already on CmdRes.

File: test/k8sT/Policies.go
Comment (raybejjani) [reply test/k8sT/Policies.go]: I switched this to not use WithTimeout.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-focus k8s.*

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-gke

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfiguration.*

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfiguration.*

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfiguration.*

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfiguration.*

**Comment by raybejjani** [comment]
test-focus K8sK8sDatapathConfig.*

**Comment by raybejjani** [comment]
test-focus K8sDatapathConfig.*

**Comment by raybejjani** [comment]
test-me-please

File: test/k8sT/DatapathConfiguration.go
Comment (joestringer) [inline test/k8sT/DatapathConfiguration.go]: monitorCancel won't be run if the `Expect(err)` fails one line above.

**Review by joestringer — CHANGES_REQUESTED** [review]
This is awesome, thanks for picking this up!

Minor nits below around calling of the cancel function.

**Review by nebril — APPROVED** [review]
Agreed with Joe's comment, aside from that LGTM!

File: test/k8sT/DatapathConfiguration.go
Comment (raybejjani) [reply test/k8sT/DatapathConfiguration.go]: Good catch. I remember thinking that I had to do it this way but I can't see why now (maybe I though it was created inside the body function? clearly it isn't #codingmysteries).

**Comment by raybejjani** [comment]
test-gke

**Review by joestringer — APPROVED** [review]
LGTM, assuming `Expect()` doesn't do anything weird that might cause the defer to be skipped.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
I tested by adding a forced fail and some prints. I see the prints on fail and pass so Expect doesn't short-circuit something to bypass defers.

I'm going to merge this since. The test failures I do see are wholly unrelated (including test timeouts).

**Pull request merged**
